### PR TITLE
Initial Commit (draft version)

### PR DESCRIPTION
**Pull request by step-stp**
Initial Commit (draft version)

---
This change is [Reviewable](https://reviewable.io/reviews/stratumn/java-chainscript/1)

### CODE REVIEW

**Comment by step-stp** [comment]
Hi Jérémie,

Please find below the feedback:

1- SignatureUtil and SignatureUtility are repetitive and will be merged during our internal code review process.

2-Yes , upon build chainscript.proto is being copied to target folder were a plugin is being used to generate source files. There seems to be a problem with the maven build and we will work towards fixing it in the next commit

Please note that above will be resolved in the next commit later this week.

Thanks,
Yasser

**Comment by bejito** [comment]
Canonicaljson should be imported rather than copy pasted in here. 
Is this expected for later ?

**Comment by step-stp** [comment]
Dear Mourad ,

Sure , we can import the Canonicaljson once you add  this library to the maven repo .

Regards ,



Kamel Abiad
Team Lead

Mobile: +961 3 941 161
Skype: kamel.imc


[cid:image001.gif@01CE5BC1.7E31ED20]<http://www.step-stp.com/>


28 Gladstonos Street, 1095 Nicosia, Cyprus
Phone +357 22 660 111 fax +357 22 451 126
----------------------------------------------------
John Kennedy Street, Mimosa Bldg, 1st Floor
Ain El-Mreisseh, Beirut 2030-709, Lebanon
Phone +961 1 366 919 ext 101 Fax +961 1 366 919

**Comment by step-stp** [comment]
Dear Mourad ,

Please find below my reply regarding the last point : (Do we need to template the interface just for return types for the methods ? It's weird to have a class templated by itself. Can't we have the methods return an ILinkBuilder instead ?)

From design perspective a builder pattern should ensure that all methods within the builder return the same data type.  If we just set the return type of the methods to ILinkBuilder,  an implementation could  have a "poisonous"  method returning a different subtype of the ILinkBuilder.  Example from the google protobuf library  com.google.protobuf.GeneratedMessageV3.Builder<BuilderType>    has the following def.
public abstract static class Builder <BuilderType extends Builder<BuilderType>>  extends AbstractMessage.Builder<BuilderType>

For the other points we will fix and commit .

Regards ,


Kamel Abiad
Team Lead

Mobile: +961 3 941 161
Skype: kamel.imc


[cid:image001.gif@01CE5BC1.7E31ED20]<http://www.step-stp.com/>


28 Gladstonos Street, 1095 Nicosia, Cyprus
Phone +357 22 660 111 fax +357 22 451 126
----------------------------------------------------
John Kennedy Street, Mimosa Bldg, 1st Floor
Ain El-Mreisseh, Beirut 2030-709, Lebanon
Phone +961 1 366 919 ext 101 Fax +961 1 366 919

File: src/main/java/com/stratumn/chainscript/Link.java [hunk 1]
```diff
@@ -0,0 +1,621 @@
...
+      }
+   }
+   
+   /***
+    * Returns custom class
+    * @param clazz
+    * @return
+    * @throws Exception
+    */
+   public <T> T data(Class<T> clazzOfT) throws Exception
```
Comment (bejito) [inline src/main/java/com/stratumn/chainscript/Link.java]: Is there a reason why we are throwing Exceptions and not ChainscriptExceptions in most of these methods ?
throws Exception is too generic to be helpful IMO, we should avoid it when possible

**Comment by step-stp** [comment]
Dear Mourad ,

Kindly note that we will commit the fix for the below in a few .

Regards ,


Kamel Abiad
Team Lead

Mobile: +961 3 941 161
Skype: kamel.imc


[cid:image001.gif@01CE5BC1.7E31ED20]<http://www.step-stp.com/>


28 Gladstonos Street, 1095 Nicosia, Cyprus
Phone +357 22 660 111 fax +357 22 451 126
----------------------------------------------------
John Kennedy Street, Mimosa Bldg, 1st Floor
Ain El-Mreisseh, Beirut 2030-709, Lebanon
Phone +961 1 366 919 ext 101 Fax +961 1 366 919

**Pull request merged**